compiler/next/include/chpl/util/hash.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UTIL_HASH_H
+#define CHPL_UTIL_HASH_H
+
+#include <cstddef>
+#include <functional>
+#include <string>
+#include <tuple>
+#include <utility>
+
+namespace chpl {
+
+// Combine two hash functions
+inline size_t hash_combine(size_t hash, size_t other) {
+  return hash ^ (other + 0x9e3779b9 + (hash << 6) + (hash >> 2));
+}
+
+// FNV-1a hash function
+#define FNV1A(PARAM, LOOP, C)                                      \
+inline size_t hash(PARAM) {                                        \
+  size_t seed = 0xcbf29ce484222325;                                \
+  for(LOOP)                                                        \
+    seed = (seed ^ static_cast<unsigned char>(C)) * 0x100000001b3; \
+  return seed;                                                     \
+}
+
+// FNV-1a hash function for null-terminated C strings
+FNV1A(const char* s, ; auto c = *s; ++s, c)
+
+// FNV-1a hash function for C++ std::string
+FNV1A(const std::string& s, auto c : s, c)
```

Comment:
I can write the functions separately, or put the FNV-1a constants into `constexpr` variables or `#define`s, so that you don't have to read the strange macro arguments.

Reply:
Thanks, I'm probably happy with many of these variations.